Handle transparncy correctly for watermark

elzekool opened the pull request:
Handle alpha transparency correctly for 24-bit PNG image

### Description
The watermark functionality uses imagecopymerge for copying the
watermark into the image, this function loses the alpha information.

Therefor use imagecopy for 100% opacity and use imagefilter in other
cases.

### Fixed Issues (if relevant)
1. magento/magento2#10661: Opacity png watermark became white box on product images

### Manual testing scenarios
1. Create a 24-bit PNG image with alpha transparency (or use the image in issue in #10661)
2. Configure the image under "Product image watermarks"
3. Leave size empty, set image position to something different than stretch
4. Create a product with a JPEG image
5. See that the watermark is applied with the correct transparency.

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

elzekool committed:
Handle transparncy correctly for watermark

The watermark functionality uses imagecopymerge for copying the
watermark into the image, this function loses the alpha information.

Therefor use imagecopy for 100% opacity and use imagefilter in other
cases.

Resolves: #10661

orlangur commented on lib/internal/Magento/Framework/Image/Adapter/Gd2.php:
```diff
@@ -465,7 +465,7 @@ public function watermark($imagePath, $positionX = 0, $positionY = 0, $opacity =
             $positionX = $this->_imageSrcWidth / 2 - imagesx($watermark) / 2;
             $positionY = $this->_imageSrcHeight / 2 - imagesy($watermark) / 2;
-            imagecopymerge(
+            $this->imagecopymergeWithAlphaFix(
```
I believe the proper name for this function should be something like `copyImage` as we are not even using `imagecopymerge` anymore.

orlangur commented on lib/internal/Magento/Framework/Image/Adapter/Gd2.php:
```diff
@@ -778,4 +778,68 @@ protected function _createEmptyImage($width, $height)
+     *
+     * @return bool
+     */
+    private function imagecopymergeWithAlphaFix(
```
Please just mention in PHPDoc of this method that signature matches http://php.net/manual/en/function.imagecopymerge.php function but name arguments properly: `$destinationImage`, `$sourceImage`, `$destinationX`, etc.

orlangur commented on lib/internal/Magento/Framework/Image/Adapter/Gd2.php:
```diff
@@ -778,4 +778,68 @@ protected function _createEmptyImage($width, $height)
+
+        $sizeX = imagesx($src_im);
+        $sizeY = imagesy($src_im);
+        if (false === $sizeX || false === $sizeY) {
```
Please convert all Yoda-style conditions to normal ones.

Is checking for `false` really necessary here - do we need to continue execution flow when `$sizeX === 0`?

orlangur commented:
Thanks, fix looks quite good in general, especially that you changed all the places. Probably we should even add `imagecopymerge` function to list of disallowed functions :)

Would be really nice to cover fix with integration test using mentioned image. Please tell us if you need any assistance with that.

Koc commented:
Have checked manually this PR on my env - works perfectly, thanx.

Does image magick also requires changes?

orlangur commented:
Hi @elzekool, are you going to continue work on this PR?

orlangur commented:
@Koc no as problematic function `imagecopymerge` belongs to `gd` extension :)

As you already tested this PR, maybe you will be interested in fixing CR notes so that it can be merged? You just need to checkout it locally as described in https://help.github.com/articles/checking-out-pull-requests-locally/, commit your changes on top and create a new PR. With merge of new PR they both will be marked as Merged.

orlangur commented:
Ping @Koc @elzekool :)

orlangur commented:
Closing due to inactivity. Feel free to reach me out anytime later if you wish to continue work on pull request and it will be reopened.

elzekool commented:
Hi, sorry for the late response. Been realy busy, want to pick it up again

elzekool committed:
Merge remote-tracking branch 'origin/2.2-develop' into bugfix/handle-alpha-transparency-correctly-for-watermark

elzekool committed:
Fix handling of watermark with alpha for Imagick

When adding a watermark that already has an alpha channel this was reset
because the overal opacity was set for the image. Instead we should
multiply the original alpha with the requested opacity

elzekool committed:
Handle non true color images with GD2

elzekool committed:
Add integration test for watermark with alpha

Added integration tests for handling alpha transparency in watermarks.
Renamed the current test for watermarks as it only tests if the
watermark is correctly places

orlangur commented:
@elzekool no problem, please go ahead 👍

elzekool committed:
Fix issues with static tests

magento-engcom-team commented:
@elzekool thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.
